File: 12865.py
n, w = map(int, input().split())
arr = []
dp = [[0 for i in range(w+1)] for j in range(n)]
for i in range(n):
    arr.append(list(map(int, input().split())))
for i in range(n):
    for j in range(1, w+1):
        if arr[i][0] > j:
            dp[i][j] = dp[i-1][j] #담을 수 없으면 내꺼 안담고 현재 무게의 가방의 최대값 들고오기
        else:
            dp[i][j] = max(dp[i-1][j], dp[i-1][j-arr[i][0]]+arr[i][1])
print(dp[n-1][w])

File: 12865.cpp
#include<iostream>
#include<vector>
#include<algorithm>
 
using namespace std;
 
int ans[100][100001];
 
int main()
{
    int N = 0;
    int can = 0;
    vector<pair<int, int> > arr;
 
    cin >> N >> can;
 
    for (int i = 0; i < N; i++){
    	int a, b;
        cin >> a >> b;
        arr.push_back(make_pair(a, b));
    }
 
    for (int i = 0; i < N; i++)    {
        for (int j = 0; j <= can; j++){
            if(i == 0)
            {
	            if(arr[i].first <= j){
	                ans[i][j] = arr[i].second;
	            	continue;
	        	}
            }
            else if(arr[i].first <= j){
            	            //직전 무게의 최대 가치, 현재 무게의 가치 + 이전에 내 무게를 뺏을 때 만큼의 가치
                ans[i][j] = max(ans[i-1][j], arr[i].second + ans[i-1][j-arr[i].first]);
            }
            else{
                ans[i][j] = ans[i-1][j]; //이전의 최대값
            }
        }
    }
    cout << ans[N-1][can];
    return 0;
}
